### DISCUSSION TITLE
coder/code-server vs nuclide vs vscode-server

### DISCUSSION THREAD

**Discussion by paulchill** [coder/code-server#5783] [Q&A] [answered]
Hey Guys,

Need some help, feeling outdated :)

I used to work at FB a long time ago and previously used Nuclide for remote dev.

Now i have a startup and dealing with algorithms and code bases which can't be downloaded outside of our dev servers.

We've been using Nuclide for remote dev for the past few years even though it's deprecated.

One thing i noticed about vscode remote ssh extension, when trying it today, it has the download option of files or folders.

- How similar is code-server to Nuclide in security? 
- Which protocols / ports are used for firewall restricted environments in AWS?
- Can file / folder downloads be restricted from the server for the client vs code?
- Can this be used with multiple users on one powerful dev server? (our current setup - each dev has a jailed home)
- What's the main difference between code-server / vscode-sever?

Thanks for the guidance guys 👍

**Comment by lg-nineoneninethreenine** [comment, marked as answer]
Well I ain't a coding expert, I am just 14 years but I could help you on some tidbits.


> How similar is code-server to Nuclide in security?

afaik code-server is built from ground up to be lightweight, easy to deploy and secure. I mean, the password you get after installing is a hash code by Argon2. So it is pretty secure :)
Tbh I have never used Nuclide, but this does my job to a satisfactory level.

> Can file / folder downloads be restricted from the server for the client vs code?

If I am not wrong, we had a request or bug or discussion on this. I think you have to check user settings

> Can this be used with multiple users on one powerful dev server? (our current setup - each dev has a jailed home)


Since code-server is made by Coder, they can(eff it, they did) integrate it into their Coder product. It is developer workspaces, but simplified. So, yeah. If the dev server is powerful, you can install Debian and set up CoderOSS on it(make sure you install Terraform separately) and provision users and admins. It would be with close-to-limited access, the public URL changes on each restart, but if you get a domain, you could redirect login there.

> Which protocols / ports are used for firewall restricted environments in AWS?

Well idk but @jsjoeio might know. If I am not wrong, it could be just port 8080.

> What's the main difference between code-server / vscode-sever?

code-server is purely open-sourced, with a huge and diverse community(in terms of the people as well as the tech they use) and a tad unofficial. That's why you can't easily hit a VSCode extension; you should use OpenVSX or the new marketplace here. But this is more lightweight. As I said above, code-server is more lightweight, and on top of that, a bit isolated - it installs into .local on my Pop!_ OS 22.04 instance. I tried vscode-server back in my Zorin days(a few months ago) and that thing was basically crap. It used the name code-server and would fail to start half the time. Serious limitations that ain't over here.

- bonus: vscode-server has a really bulky requirement if I can remember properly. code-server just needs POSIX, 1GHz core and a gig of RAM.

POSIX - Macs and Linux PCs :wink:

PS: curse you gboard and GitHub mobile. It ain't as easy as I thought!

**Reply by lg-nineoneninethreenine** [reply]
Dang, that was _huge_! Sorry!

**Reply by paulchill** [reply]
Thanks so much for taking the time for the detailed overview :)

**Reply by lg-nineoneninethreenine** [reply]
No worries :) this makes me feel liberated from my real life tbh. I'm happy to have helped you!

PS: for in-depth answers, @jsjoeio could help out. I could be wrong here and there since I am here since perhaps a year, and I am not well-versed with coding, I just told all that from my experience. If you need better answers/explanations, the maintainers could give the best answers

**Reply by jsjoeio** [reply]
> Which protocols / ports are used for firewall restricted environments in AWS?

Great question! Thinking out loud here...so there's HTTP/HTTPS, a couple WSS connections and I think that's it...For ports, default is 8080 for running it but not sure about other ports  🤔 @code-asher

**Reply by jsjoeio** [reply]
Also @im-coder-lg you are so awesome for writing such a long thru response! Thank you!

**Reply by lg-nineoneninethreenine** [reply]
No worries! That was refreshing after a day of torture from the bozos at school! I hope the answers were correct though.

**Reply by code-asher** [reply]
Great answers.

> How similar is code-server to Nuclide in security?

I am also not familiar with Nuclide but I can say security-wise code-server has the basics you would expect of most websites.  Cookie-based token authentication, CSP, rate limiting (although I believe rate limiting is unfortunately missing on some endpoints).  In Coder we disable the built-in authentication and run an authentication wrapper around it that supports oauth.

code-server is basically a remote shell so once you are in you can do anything on the machine the user has permissions to do.  The way to do multi-tenancy would be to spawn a code-server using each separate user or do some sort of Docker/Kubernetes setup and give each user their own container.  Sounds like the jailed home thing you have going on would be perfect.

There is also a built-in proxy which currently cannot be disabled.  It can be used by authenticated users to access any port on the machine by visiting the `/proxy/{port number}/` endpoint.  We might gain an option to disable it in https://github.com/coder/code-server/issues/5826.

> Can file / folder downloads be restricted from the server for the client vs code?

If I recall correctly you can pass `--disable-file-downloads` or use the environment variable `CS_DISABLE_FILE_DOWNLOADS=true`.  It blocks the most convenient path but it may still possible to work around it for example by running a file browser via the proxy, directly accessing the endpoints code-server itself uses to load file contents, good old copy and paste, or sending the contents somewhere by running some command in code-server's terminal.  I think most of these can be blocked off in due time but for now they remain open.

> Which protocols / ports are used for firewall restricted environments in AWS?

Can confirm that code-server only uses one port at the moment (8080 by default as noted).  It can also be ran on a socket which may be preferable in a multi-user environment.

As noted currently the only protocol in use is HTTP/S.  There is some talk about supporting arbitrary protocols for the proxy (for example to proxy Postgres). This would run on a separate port.  https://github.com/coder/code-server/issues/5723
 
> code-server vs vscode-server

Both code-server and vscode-server wrap OSS Code but as noted vscode-server is able to use proprietary bits that code-server cannot and code-server has a set of patches to Code that allow self-hosting the server as well as a few bug fixes that have not been upstreamed yet.  Those differences can all be seen here in the patch descriptions: https://github.com/coder/code-server/tree/main/patches

As far as the wrappers go I would say the main differences are that we add a different implementation of token-based authentication, the proxy, a /healthz endpoint, and a few deprecated extras like `--link`.
